sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/metadata/datasource/dialect/H2DataSourceMetaData.java

```diff
@@ -31,24 +37,43 @@
  */
 @Getter
 public final class H2DataSourceMetaData implements MemorizedDataSourceMetaData {
-    
+
     private static final int DEFAULT_PORT = -1;
-    
+
     private final String hostName;
-    
+
     private final int port;
-    
+
     private final String schemaName;
-    
+
+    private final String catalog;
+
     private final Pattern pattern = Pattern.compile("jdbc:h2:(mem|~)[:/]([\\w\\-]+);?\\S*", Pattern.CASE_INSENSITIVE);
-    
+
     public H2DataSourceMetaData(final String url) {
         Matcher matcher = pattern.matcher(url);
         if (!matcher.find()) {
             throw new UnrecognizedDatabaseURLException(url, pattern.pattern());
         }
         hostName = "";
         port = DEFAULT_PORT;
-        schemaName = matcher.group(2);
+        catalog = matcher.group(2);
+        schemaName = null;
+    }
+
+    public H2DataSourceMetaData(final DataSource dataSource) {
```

Comment:
Why do we need this funciton? How about passing `dataSourceMetadata` containing `jdbcUrl` and `username` to constructor above?

Reply:
At present we only need jdbcurl and username ,but at future ,maybe we need more datasouce infomation,so,it is possibility that will use this datasource to get more infomation.